### PR TITLE
UI: Harvesting message includes fleet remaining free space

### PR DESCRIPTION
**Pull request by SomeTroglodyte**
**Feature**

This PR addresses the feature described in issue #10228

## Summary
As the issue suggested - my skills are very limited, but finding the message was trivial, and adapting the "which ships should I count" loop from Engine::Place seemed straightforward too.

## Screenshots
<details>

Sorry for the size... Also not vanilla rules, plugins active. But for a simple message...

image

</details>

## Testing Done
Took screenshot. :zany_face: 
Actually, with my mods I can generate way more flotsam in a very short time than one can in vanilla, so it got some stress test.

## Save File
May add one later, or on request - should be vanilla and not require any plugin

## Performance Impact
A loop to get a total is more expensive than the one method call this message had before, but -as mentioned- my tests suggest that has little impact, especially as this does not repeat per frame, it's only once per flotsam?

### CODE REVIEW

**Review by TheGiraffe3 — APPROVED** [review]
Looks like it _should_ work. I’ll ask @tibetiroka though as someone who can code.

File: source/Engine.cpp [hunk 1]
```diff
@@ -2441,12 +2441,20 @@ void Engine::DoCollection(Flotsam &flotsam)
 	}
 
 	// Unless something went wrong while forming the message, display it.
-	if(!message.empty())
-	{
-		int free = collector->Cargo().Free();
+	if(message.empty()) return;
```
Comment (TomGoodIdea) [inline source/Engine.cpp:2444]: style
```suggestion
	if(message.empty())
		return;
```

File: source/Engine.cpp [hunk 2]
```diff
@@ -2441,12 +2441,20 @@ void Engine::DoCollection(Flotsam &flotsam)
...
+
+	int free = collector->Cargo().Free();
+	int total = 0;
+	for(const shared_ptr <Ship> &ship : player.Ships())
```
Comment (TomGoodIdea) [inline source/Engine.cpp:2448]: style
```suggestion
	for(const shared_ptr<Ship> &ship : player.Ships())
```

File: source/Engine.cpp [hunk 3]
```diff
@@ -2441,12 +2441,20 @@ void Engine::DoCollection(Flotsam &flotsam)
...
+		if(!ship->IsParked() && ship->GetSystem() == player.GetSystem())
```
Comment (TomGoodIdea) [inline source/Engine.cpp:2449]: Do we want to include disabled ships?

File: source/Engine.cpp [hunk 4]
```diff
@@ -2441,12 +2441,20 @@ void Engine::DoCollection(Flotsam &flotsam)
...
+			total += ship->Cargo().Free();
+
+	if(free == total)
 		message += " (" + Format::CargoString(free, "free space") + " remaining.)";
```
Comment (TomGoodIdea) [inline source/Engine.cpp:2453]: Maybe move this line (and a similar one from the `else` block) outside the `if..else` to avoid duplicating code?

Comment (TomGoodIdea) [reply source/Engine.cpp:2453]: Both options are fine for me, but probably some people would say the second one is clearer.

**Review by TomGoodIdea — APPROVED** [review]
Perhaps we could sum the free cargo space as double and then cast to int to increase accuracy, but I think it's fine as is.

File: source/Engine.cpp (hunk 3 above)
Comment (warp-core) [reply source/Engine.cpp:2450]: I think I'll leave it as is for now and see if people find that they would prefer disabled ships not be counted or not.

**Pull request merged**